### PR TITLE
Category Encoding Layer

### PR DESCRIPTION
**Pull request by AdamLang96**
* Category Encoding Preprocessing Layer Co-authored-by:
David Kim (@koyykdy) <dok098@ucsd.edu>
Brian Zheng (@brianzheng123) <brianzheng345@gmail.com>

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

- - -
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/6855)

### CODE REVIEW

**Comment by AdamLang96** [comment]
@mattsoulanille I was hoping to get a review of this whenever you are available to. Thank you!

File: tfjs-layers/src/layers/preprocessing/category_encoding_test.ts [hunk 1]
```diff
@@ -0,0 +1,141 @@
+import { describeMathCPUAndGPU, expectTensorsClose} from '../../utils/test_utils';
+import { Tensor, tensor} from '@tensorflow/tfjs-core';
+import { CategoryEncoding } from './category_encoding';
+import * as utils from './preprocessing_utils';
+
+describeMathCPUAndGPU('Layer Output', () => {
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/category_encoding_test.ts:6]: Include the name of the layer in the `describe`, e.g.

```suggestion
describeMathCPUAndGPU('Category Encoding', () => {
```

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+  inputs = getExactlyOneTensor(inputs);
+
+  if(inputs.dtype !== 'int32') {
+        inputs = K.cast(inputs, 'int32');
+    }
+
+  if(outputMode === int) {
+    return inputs;
+  }
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:27]: I don't see `int` output mode ever passed to this function, and I can't find it in the [Keras docs for category encoding](https://www.tensorflow.org/api_docs/python/tf/keras/layers/CategoryEncoding). What is it for?

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+   } else {
+
+    binCounts = denseBincount(denseBincountInput, [], depth, binaryOutput);
+
+   }
+
+  if(outputMode !== tfIdf) {
+    return binCounts;
+  }
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:73]: What is `tfIdf` for? I don't see it in these [keras docs](https://www.tensorflow.org/api_docs/python/tf/keras/layers/CategoryEncoding).

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+export const count = 'count';
+export const tfIdf = 'tfIdf';
+
+export function encodeCategoricalInputs(inputs: Tensor|Tensor[],
+                                        outputMode: string,
+                                        depth: number,
+                                        weights?: Tensor1D|Tensor2D|TensorLike):
+                                        Tensor|Tensor[] {
+
+  inputs = getExactlyOneTensor(inputs);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:19]: Nit: Avoid reassigning `inputs` from the function's arguments to make this easier to read.
```suggestion
  let input = getExactlyOneTensor(inputs);
```

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+  if(inputs.rank > 2) {
+    throw new ValueError(`When outputMode is not 'int', maximum output rank is 2
+    Received outputMode ${outputMode} and input shape ${originalShape}
+    which would result in output rank ${inputs.rank}.`);
+  }
+
+  const binaryOutput = [multiHot, oneHot].includes(outputMode);
+
+  let denseBincountInput;
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:49]: Give this a type
```suggestion
  let denseBincountInput: Tensor1D | Tensor2D;
```

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+  if(inputs.rank === 1) {
+    denseBincountInput = inputs as Tensor1D;
+  }
+
+  if(inputs.rank === 2) {
+    denseBincountInput = inputs as Tensor2D;
+  }
+
+  let binCounts;
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:59]: Give this a type. I think it's this, but not sure.
```suggestion
  let binCounts: Tensor1D | Tensor2D;
```

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+  if(weights === null || weights === undefined) {
+    throw new ValueError(
+      `When outputMode is 'tfIdf', weights must be provided.`
+      );
+  } else {
+
+    return mul(binCounts, weights);
+
+  }
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:83]: Nit: `weights` has type `Tensor1D | Tensor2D | TensorLike | undefined`. All the types in this union with `Tensor` in them are truthy (`Boolean()` of them is true), so you can write this instead:

```suggestion
  if (weights) {
    return mul(binCounts, weights);
  } else {
      throw new ValueError(
        `When outputMode is 'tfIdf', weights must be provided.`
      );
  }
```

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
Thanks for the PR! I have a few changes for you to take a look at. Feel free to @ me immediately in future PRs, or when you're ready for me to take another look, and I'll respond sooner.

**Comment by AdamLang96** [comment]
@mattsoulanille Thank you for the review! I'll ping you once these changes have been implemented

**Comment by AdamLang96** [comment]
hi @mattsoulanille I made the requested changes and checked that the linter was passing via `yarn lint` however I am failing the CI build. Do you have any insight on why this could be happening?

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+import { ValueError } from '../../errors';
+import * as K from '../../backend/tfjs_backend';
+
+export const int = 'int';
+export const oneHot = 'oneHot';
+export const multiHot = 'multiHot';
+export const count = 'count';
+export const tfIdf = 'tfIdf';
+
+export function encodeCategoricalInputs(inputs: Tensor|Tensor[],
```
[the remark this answers is not in the document]
Comment (AdamLang96) [reply tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:19]: I am porting over the logic and file structure from Keras. in [category_encoding.py](https://github.com/keras-team/keras/blob/master/keras/layers/preprocessing/category_encoding.py) they layer calls `encode_categorical_inputs` which lives in [preprocessing_utils.py](https://github.com/keras-team/keras/blob/master/keras/layers/preprocessing/preprocessing_utils.py)

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts (hunk 3 above)
Comment (AdamLang96) [reply tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts]: `tfIdf` is an encoding algorithm implemented in [encode_categorical_inputs](https://github.com/keras-team/keras/blob/40c54a663bc159b5629206580fa15d2c0af343fe/keras/layers/preprocessing/preprocessing_utils.py#L96)

**Comment by mattsoulanille** [comment]
Looks like it's failing one of the bounds tests:
```
Chrome 105.0.0.0 (Mac OS 10.13.6) Category Encoding webgl2 {"WEBGL_VERSION":2,"WEBGL_CPU_FORWARD":false,"WEBGL_SIZE_UPLOAD_UNIFORM":0} Raises Value Error if max input value !<= numTokens FAILED
	Expected function to throw an exception with message 'Input values must be between 0 < values <= numTokens
	        with numTokens=3', but it threw an exception with message 'Input values must be between 0 < values <= numTokens
	          with numTokens=3'.
	    at <Jasmine>
	    at UserContext.<anonymous> (tfjs-layers/src/layers/preprocessing/category_encoding_test.ts:139:6 <- tfjs/tfjs-layers/src/tfjs-layers_test_bundle.js:55301:59)
	    at <Jasmine>
```
This is strange; the two error messages look identical to me (unless I've just completely missed something).

Same for output mode tests:
```
Chrome 105.0.0.0 (Mac OS 10.13.6) Tests for preprocessing utils cpu {} Thows an error if input rank > 2 FAILED
	Expected function to throw an exception with message 'When outputMode is not 'int', maximum output rank is 2
	    Received outputMode 'multiHot' and input shape 2,2,1
	    which would result in output rank 3.', but it threw an exception with message 'When outputMode is not 'int', maximum output rank is 2
	    Received outputMode multiHot and input shape 2,2,1
	    which would result in output rank 3.'.
	    at <Jasmine>
	    at UserContext.<anonymous> (tfjs-layers/src/layers/preprocessing/preprocessing_utils_test.ts:60:6 <- tfjs/tfjs-layers/src/tfjs-layers_test_bundle.js:55403:72)
	    at <Jasmine>
```
The failures are reproducible locally with `yarn bazel run //tfjs-layers:tfjs-layers_webgl2_test` (run in the root of the repository). You can focus the test that's failing by replacing `it` with `fit`, or you can use this url (gotten by clicking `DEBUG` in Chrome karma, waiting for the tests to finish, and then clicking the broken one in the list of tests), although you might need to change the port.

[Full logs for layers tests](https://source.cloud.google.com/results/invocations/09dd8ba2-a93f-440d-a587-2b1003f14fb9/targets/%2F%2Ftfjs-layers:bs_chrome_mac_tfjs-layers_webgl2_test/log)

**Comment by mattsoulanille** [comment]
Oh, I see the issue. JS does not let you wrap strings by just adding a newline in them. Well, it does, but it includes the newline and all the indentation spaces. You'll need to split those strings into multiple different strings that you add together. I'll add an example.

File: tfjs-layers/src/layers/preprocessing/category_encoding.ts [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+        const greaterEqualMax = greater(this.numTokens, maxValue)
+                                                    .bufferSync().get(0);
+
+        const greaterMin = greaterEqual(minValue, 0).bufferSync().get(0);
+
+        if(!(greaterEqualMax && greaterMin)) {
+
+          throw new ValueError(
+          `Input values must be between 0 < values <= numTokens
+          with numTokens=${this.numTokens}`);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/category_encoding.ts:108]:
```suggestion
          throw new ValueError('Input values must be between 0 < values <='
            + ` numTokens with numTokens=${this.numTokens}`);
```

File: tfjs-layers/src/layers/preprocessing/category_encoding_test.ts [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+  });
+
+  it('Raises Value Error if max input value !<= numTokens', () => {
+    const categoryData   = tensor([7, 2, 0, 1]);
+    const numTokens = 3;
+    const encodingLayer = new CategoryEncoding({numTokens,
+                                        outputMode: 'multiHot'});
+    expect(() => encodingLayer.apply(categoryData))
+    .toThrowError(`Input values must be between 0 < values <= numTokens
+        with numTokens=${numTokens}`);
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/category_encoding_test.ts:140]:
```suggestion
    .toThrowError('Input values must be between 0 < values <= numTokens'
      + ` with numTokens=${numTokens}`);
```

**Review by mattsoulanille — CHANGES_REQUESTED** [review]
Quick review showing how to fix the multiline string errors. I'll do a full review in a bit.

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (mattsoulanille) [reply tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:19]: Sounds good. Thanks for the explanation!

File: tfjs-layers/src/layers/preprocessing/category_encoding.ts [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+              `countWeights is not used when outputMode !== count.
+              Received countWeights=${kwargs['countWeights']}`);
+          }
+          const countWeightsArg = getExactlyOneTensor(kwargs['countWeights']);
+
+          if(countWeightsArg.rank === 1) {
+            countWeights = countWeightsArg as Tensor1D;
+          } if(countWeightsArg.rank === 2) {
+            countWeights = countWeightsArg as Tensor2D;
+            }
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/category_encoding.ts:94]: Nit: You're assigning to a variable of type `Tensor1D | Tensor2D`, so the extra information you get when checking if `countWeightsArg` is one or two-dimensional is immediately lost.

Instead, you can just assign directly:
```suggestion
          if (countWeightsArg.rank === 1 || countWeightsArg.rank === 2) {
            const countWeights = getExactlyOneTensor(kwargs['countWeights']) as Tensor1D | Tensor2D;
          }
```

You can remove the if statement if you already know countWeightsArg is of rank 1 or 2.

File: tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+
+  let denseBincountInput: Tensor1D | Tensor2D;
+
+  if(input.rank === 1) {
+    denseBincountInput = input as Tensor1D;
+  }
+
+  if(input.rank === 2) {
+    denseBincountInput = input as Tensor2D;
+  }
```
Comment (mattsoulanille) [inline tfjs-layers/src/layers/preprocessing/preprocessing_utils.ts:63]: Nit: Same here as in countWeights.

```suggestion
  const denseBincountInput = input as Tensor1D | Tensor2D;
```

**Review by mattsoulanille — APPROVED** [review]
LGTM with a few nits. Thanks!

**Review by ahmedsabie — APPROVED** [review]

**Comment by AdamLang96** [comment]
@mattsoulanille I just updated the branch and it caused the build to fail. Please let me know if there is something I can fix. Thanks for all of the help!

**Comment by mattsoulanille** [comment]
@AdamLang96 We just changed some GCP Cloudbuild permissions and they accidentally broke CI for a few minutes. Should be fixed now, so I'll re-run the tests.

**Pull request merged**